Add the Enroll Kibana API

jkakavas opened the pull request:
This change adds the Enroll Kibana API that allows a kibana instance to
configure itself to communicate with a secured elasticsearch cluster

jkakavas committed:
Add the Enroll Client API

This change adds the Enroll Client API that allows a client to
configure itself to communicate with a secured elasticsearch cluster

elasticmachine commented:
Pinging @elastic/es-security (Team:Security)

jkakavas commented:
CC @azasypkin

jkakavas commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/action/TransportClientEnrollmentAction.java:
```diff
@@ -0,0 +1,112 @@
+    @Override protected void doExecute(
+        Task task, ClientEnrollmentRequest request, ActionListener<ClientEnrollmentResponse> listener) {
+        try {
+            final String httpCaCert = "httpCa.crt";
```
We're considering options for how generic we want to be in this and might update the way we figure out the certificate.

jkakavas committed:
fix doc test skip

jkakavas committed:
fix typos

jkakavas committed:
remove unused import

jkakavas committed:
fix logger usage

jkakavas committed:
fix validation and tests

jkakavas committed:
fix test and add api as operator only

jkakavas committed:
Merge remote-tracking branch 'origin/master' into enroll-client-api

jkakavas committed:
Move API to the security plugin

jkakavas committed:
Read HTTP CA and key from they keystore used for the HTTP layer

jkakavas committed:
Change API registered endpoint

jkakavas committed:
fixes

jkakavas committed:
Merge remote-tracking branch 'origin/master' into enroll-client-api

jkakavas committed:
Disable tests that fail

jkakavas committed:
revert unwanted change

jkakavas committed:
Merge remote-tracking branch 'origin/master' into enroll-client-api

jkakavas commented on client/rest-high-level/src/main/java/org/elasticsearch/client/ClusterClient.java:
```diff
@@ -273,4 +273,5 @@ public Cancellable existsComponentTemplateAsync(ComponentTemplatesExistRequest c
         return restHighLevelClient.performRequestAsync(componentTemplatesRequest,
             ClusterRequestConverters::componentTemplatesExist, options, RestHighLevelClient::convertExistsResponse, listener, emptySet());
     }
+
```
will clean up

sethmlarson commented on rest-api-spec/src/main/resources/rest-api-spec/api/security.enroll_client.json:
```diff
@@ -0,0 +1,24 @@
+{
+  "security.enroll_client":{
+    "documentation":{
+      "url":"https://www.elastic.co/guide/en/elasticsearch/reference/master/security-aoi-enroll-client.html",
```
Typo in the URL, `aoi` -> `api`

sethmlarson commented on rest-api-spec/src/main/resources/rest-api-spec/api/security.enroll_client.json:
```diff
@@ -0,0 +1,24 @@
+          ]
+        }
+      ]
+    }
```
The API description requires a `"body": {"description": "..."}` section to signal to clients that a body is required.

jkakavas committed:
address feedback

BigPandaToo commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportClientEnrollmentAction.java:
```diff
@@ -0,0 +1,128 @@
+                    }
+                    if (shouldSetPassword(request)) {
+                        final String user = userPasswordToSet(request);
+                        assert user != null;
```
Should we throw an exception here?

jkakavas replied:
I added an assert here as this should _not_ happen, but wanted to check eitherway and for future proofing. `shouldSetPassword` should only return true when the client type is kibana and when the client type is kibana the `userPasswordToSet` is hardcoded to return `kibana_system`.

BigPandaToo reviewed (commented):
Added few comments, mostly nit. Feel free to merge after you addressing them

BigPandaToo commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/operator/OperatorOnlyRegistry.java:
```diff
@@ -35,7 +36,8 @@
         "cluster:admin/repository/analyze/blob",
-        "cluster:admin/repository/analyze/blob/read"
+        "cluster:admin/repository/analyze/blob/read",
+        ClientEnrollmentAction.NAME
```
Just seen Albert's comment here for Node Enrollment action. May be applicabble here as well.

jkakavas replied:
Yeap, see my comment on your PR too. Will update

jkakavas commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackPlugin.java:
```diff
@@ -244,12 +244,12 @@ private static boolean alreadyContainsXPackCustomMetadata(ClusterState clusterSt
             metadata.custom(TransformMetadata.TYPE) != null;
     }
-    
+
```
There is no extra line here, I simply removed some whitespace

jkakavas commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/enrollment/ClientEnrollmentAction.java:
```diff
@@ -0,0 +1,20 @@
+
+public final class ClientEnrollmentAction extends ActionType<ClientEnrollmentResponse> {
+
+    public static final String NAME = "cluster:admin/xpack/enroll/client";
```
Yeap, not only for consistency but also so that the privilege matchers apply in the same way. Given where that discussion currently is, I'll change this to `cluster:admin/xpack/security/enrollment/enroll/client`

jkakavas committed:
address feedback

jkakavas committed:
Adjusted HLRC docs

jkakavas committed:
Merge remote-tracking branch 'origin/master' into enroll-client-api

albertzaharovits commented:
@jkakavas I have some doubts that I need to clear before I'm confident approving this.

In principle, the clients get an enrollment token that contains a credential and a certificate fingerprint, which they then use to **generate** another credential as well as the certificate associated to the fingerprint.

I understand why this is done, because we cannot store raw credentials and because kibana authn with users and only api keys can expire.

But still, the fact that you can enroll multiple kibana instances and that each can set different passwords, looks to me like a brittle solution. I'll mull it over a bit more.

jkakavas commented:
> In principle, the clients get an enrollment token that contains a credential and a certificate fingerprint, which they then use to generate another credential as well as the certificate associated to the fingerprint.

> I understand why this is done, because we cannot store raw credentials and because kibana authn with users and only api keys can expire.

Also we don't want to send out kibana system user credentials in an enrollment token ( not sure this is waht you meant with "raw credentials" below. 


> But still, the fact that you can enroll multiple kibana instances and that each can set different passwords, looks to me like a brittle solution. I'll mull it over a bit more.

We can guard against this if we feel we need to and we don't want the kibana administrators to be responsible for this. ( only send the password for the first kibana instance client enroll call ). There is always the chace that kibana will switch to use service accounts before 8.0 and we'd need to restructure this API eitherway.

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/enrollment/RestClientEnrollmentAction.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    @Override public List<Route> routes() {
+        return List.of(new Route(RestRequest.Method.POST, "/_security/enroll_client"));
```
I don't love this URL paths.
See https://github.com/elastic/elasticsearch/pull/72186#discussion_r641229443

elasticmachine commented:
Pinging @elastic/clients-team (Team:Clients)

jkakavas committed:
Merge remote-tracking branch 'origin/master' into enroll-client-api

jkakavas committed:
minor fixes

jkakavas committed:
renames

jkakavas committed:
Use SecureString for the password of the kibana_system user

jkakavas committed:
remove blank lines

jkakavas committed:
remove unused files

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportKibanaEnrollmentAction.java:
```diff
@@ -0,0 +1,148 @@
+        ActionFilters actionFilters) {
+        super(KibanaEnrollmentAction.NAME, transportService, actionFilters, KibanaEnrollmentRequest::new);
+        this.environment = environment;
+        // Should we use a specific origin for this ? Are we satisfied with the auditability of the change password request as-is ?
```
cc @albertzaharovits

jkakavas replied:
I'll move on with the security origin and we can circle back then

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportKibanaEnrollmentAction.java:
```diff
@@ -0,0 +1,148 @@
+                for (NodeInfo nodeInfo : nodesInfoResponse.getNodes()) {
+                    nodeList.add(nodeInfo.getInfo(HttpInfo.class).getAddress().publishAddress().toString());
+                }
+                // Should we guard against setting the password of the kibana_system user multiple times?
```
Until we get an update that service accounts can be used in Kibana, I'm working with the assumption that we'll use a builtin user

jkakavas commented:
@albertzaharovits , @BigPandaToo this is ready for a review

jkakavas committed:
Adjust enroll node API endpoint and action name

jkakavas committed:
fix action name

jkakavas committed:
Merge remote-tracking branch 'origin/master' into enroll-client-api

jkakavas committed:
Fix test

jkakavas committed:
Add instructions for generating keystores

jkakavas committed:
Don't return duplicate certificates

jkakavas committed:
I never learn

jkakavas committed:
Merge remote-tracking branch 'origin/master' into enroll-client-api

jkakavas committed:
fix typo

jkakavas committed:
fix asciidoc

BigPandaToo reviewed (changes requested):
Left few comments and suggestions

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportKibanaEnrollmentAction.java:
```diff
@@ -0,0 +1,147 @@
+        for (int i = 0; i < passwordLength; ++i) {
+            characters[i] = passwordChars[secureRandom.nextInt(passwordChars.length)];
+        }
+        return characters;
```
Can you explain what the gain would be in this case ? I'm zero filling the char array in the finally clause

BigPandaToo replied:
Just 1) reads better, 2) if sombody remove/modify the `finally` in future it may be left unnoticed. What are the benefits of doing it "manually"?

jkakavas commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/action/enrollment/TransportKibanaEnrollmentActionTests.java:
```diff
@@ -0,0 +1,169 @@
+        assertThat(changePasswordRequests.size(), equalTo(1));
+        assertThat(nodesInfoRequests.size(), equalTo(1));
+
+        assertWarnings("[keystore.password] setting was deprecated in Elasticsearch and will be removed in a future release! " +
```
Yes we needed to.  This is not part of this test, as in we're not testing something with this assertion but without assertWarnings there , the test fails. I'll replace the setup to use secure settings so that this is not needed any more

jkakavas committed:
address feedback

jkakavas committed:
Merge remote-tracking branch 'origin/master' into enroll-client-api

jkakavas commented on client/rest-high-level/src/test/java/org/elasticsearch/client/SecurityIT.java:
```diff
@@ -169,6 +170,17 @@ public void testEnrollNode() throws Exception {
+            , endsWith("OWFyeGNmcwovSDJReE1tSG1leXJRaWxYbXJPdk9PUDFTNGRrSTFXbFJLOFdaN3c9Ci0tLS0tRU5EIENFUlRJRklDQVRFLS0tLS0K"));
+        List<String> nodesAddresses2 = kibanaResponse.getNodesAddresses();
+        assertThat(nodesAddresses2.size(), equalTo(1));
+    }
```
I think that it is, unless you explicitly configure the test with an external cluster or certain nodes, types

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportKibanaEnrollmentAction.java:
```diff
@@ -0,0 +1,148 @@
+                try {
+                    client.execute(ChangePasswordAction.INSTANCE, changePasswordRequest, ActionListener.wrap(response -> {
+                        logger.debug("Successfully set the password for user [kibana_system] during kibana enrollment");
+                        listener.onResponse(new KibanaEnrollmentResponse(new SecureString(password), httpCa, nodeList));
```
@azasypkin @legrego 
What is your opinion / preference regarding getting the CA certificate ( that you already know the fingerprint of) explicitly as a field in the response vs reading it from the TLS handshake ? I thought the former is easier/preferable but this is mostly up to the consumer ( Kibana ), so I'm happy to change as needed

azasypkin replied:
> I thought the former is easier/preferable

Yeah, it'd be my preference as well. The logic to extract all necessary data from a single place (response) would certainly be easier to maintain and test.

legrego replied:
+1 to getting this from the es response instead of parsing it out of the TLS handshake

jkakavas replied:
Thanks for weighing in folks

albertzaharovits replied:
Kibana already needs to extract the certificate from the TLS handshake in order to do the authentication using the fingerprint from the enrollment token.
If ES duplicates the certificate in the response you would need to code and test for the case where we return a different certificate than we use in TLS.

azasypkin replied:
Did we consider updating our ES client libs to at least handle fingerprint matching during TLS handshake for us? I naively (and probably mistakenly) assumed that client-type-agnostic "pre-flight" things like this one can be done inside our client libs,  but I admit my knowledge about planned changes in ES client libs is very limited.

Kibana connects to Elasticsearch via `elasticsearch-js` client that, as far as I'm aware, doesn't expose the raw connection socket, so we might need to make a separate "pre-flight" handshake with Elasticsearch host outside of `elasticsearch-js` client just for the sake of fingerprint matching. But even in this case I'd still prefer to have a CA certificate in the response body.

jkakavas replied:
> Did we consider updating our ES client libs to at least handle fingerprint matching during TLS handshake for us? 

Yes this is the plan. There will be a new setting that one can set the fingerpint as a value for and the client will do the verification on TLS handshake.

azasypkin replied:
It's great to know, thanks!

jkakavas replied:
After discussing this, we realized that we do not need to return the other node addresses, at least for now. If we ever have need for this to support HA setups, we can revisit the decision

BigPandaToo reviewed (approved):
LGTM

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportKibanaEnrollmentAction.java:
```diff
@@ -0,0 +1,148 @@
+                }
+                // Should we guard against setting the password of the kibana_system user multiple times?
+                final char[] password = generateKibanaSystemPassword();
+                final ChangePasswordRequest changePasswordRequest =
```
Change password rather than put user, unnecessary assumes that the `kibana_system` user has already been created.

jkakavas replied:
Do we plan on changing this in 8.0 ? Don't we always assume that our built-in users are created ? ( setup passwords makes the same assumption )

jkakavas replied:
We discussed this live and will use change password

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportKibanaEnrollmentAction.java:
```diff
@@ -0,0 +1,148 @@
+            client.execute(NodesInfoAction.INSTANCE, nodesInfoRequest, ActionListener.wrap(nodesInfoResponse -> {
+                final List<String> nodeList = new ArrayList<>();
+                for (NodeInfo nodeInfo : nodesInfoResponse.getNodes()) {
+                    nodeList.add(nodeInfo.getInfo(HttpInfo.class).getAddress().publishAddress().toString());
```
Publish address only exposes a single address (per node). Hopefully the one reachable from Kibana (but no way to tell for sure).

It is safer to show all the addresses, but that is not possible if we must also support binding to `0.0.0.0` (ie instead of showing publish addresses we can show bind addresses, but bind can be `0.0.0.0` which is not routable).

So there is no perfect solution. I would choose to show bind addresses because the setup script won't use `0.0.0.0`. Your call.

jkakavas replied:
Good catch. I didn't think this through as much as I should have

> I would choose to show bind addresses because the setup script won't use 0.0.0.0

The only problem with this is docker where we bind on 0.0.0.0 actually. I 'll think this through on Monday and get back with a suggestion

jkakavas replied:
Thought this through a little more:

- Kibana already can communicate with an ES node ( it just sent us a request ). The idea here is that we return the http addresses for other nodes so that it might connect to other nodes if it wants/needs. 
- bind_address is always 0.0.0.0 in docker

I think publish_address is an appropriate compromise for this. We can, however, get bind_address and return those unless it is 0.0.0.0, when we would fallback to publish_address. 

@azasypkin , @legrego ping for visibility and feedback.

legrego replied:
It sounds like publish address is what we would want here. While there are no guarantees that Kibana will be able to reach ES at that address, [the docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-network.html#modules-network-binding-publishing) claim that the publish address must be reachable by all clients (if I understand correctly).

I'm not opposed to also including the bind address in the response if it's not `0.0.0.0`, however.

azasypkin replied:
Same here, returning of `publish_address` or `publish_address` + `bind_address[]` (excluding `0.0.0.0`) sounds like a reasonable approach to me.

albertzaharovits reviewed (approved):
I've left a couple of points. They need to be addressed before merging.
Nonetheless I have approved. Do ping me if there's anything I should clarify.

jkakavas commented on client/rest-high-level/src/test/java/org/elasticsearch/client/security/ClientErnollmentResponseTests.java:
```diff
@@ -0,0 +1,78 @@
+
+import static org.hamcrest.Matchers.equalTo;
+
+public class ClientErnollmentResponseTests extends ESTestCase {
```
No matter how many renames one makes, there is always one more to be done...

jkakavas commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfig.java:
```diff
@@ -127,6 +129,24 @@ X509ExtendedTrustManager createTrustManager(@Nullable Environment environment) {
+        final Enumeration<String> aliases = trustStore.aliases();
+        while (aliases.hasMoreElements()) {
+            String alias = aliases.nextElement();
+            final Certificate certificate = trustStore.getCertificate(alias);
```
I think we need `getCertificate` actually. Even if we're iterating over an entry of a <private key, certificate chain>, for instance for the node key material, what we're interested in is the leaf certificate.  Presumably the entry we put in for the CA key and certificate only has a single certificate too and not a chain. 

Can you elaborate what you were thinking?

jkakavas replied:
We discussed that these are equivalent for our use case, will keep existing method

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportKibanaEnrollmentAction.java:
```diff
@@ -0,0 +1,148 @@
+        this.environment = environment;
+        // Should we use a specific origin for this ? Are we satisfied with the auditability of the change password request as-is ?
+        this.client = new OriginSettingClient(client, SECURITY_ORIGIN);
+        this.settings = settings;
```
Makes sense!

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportKibanaEnrollmentAction.java:
```diff
@@ -0,0 +1,148 @@
+                        listener.onResponse(new KibanaEnrollmentResponse(new SecureString(password), httpCa, nodeList));
+                    }, e -> listener.onFailure(new ElasticsearchException("Failed to set the password for user [kibana_system]", e))));
+                } finally {
+                    Arrays.fill(password, '\0');
```
I see.. I'll change this to use a SecureString as Lyudmila also originally suggested, and I won't have to manually zero fill this

jkakavas replied:
We don't in other cases and I think it is ok, but I'm open to suggestions for ways to handle this that I'm missing

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/enrollment/RestKibanaEnrollmentAction.java:
```diff
@@ -0,0 +1,61 @@
+     * @param settings the node's settings
+     * @param licenseState the license state that will be used to determine if security is licensed
+     */
+    public RestKibanaEnrollmentAction(Settings settings, XPackLicenseState licenseState) {
```
will do, no worries

jkakavas committed:
partially address feedback

lockewritesdocs commented on x-pack/docs/en/rest-api/security/enroll-kibana.asciidoc:
```diff
@@ -0,0 +1,51 @@
+}
+--------------------------------------------------
+<1> The password for the `kibana_system` user
+<2> The CA certificate that is used to sign the certificates that {es}
+nodes use for TLS on the HTTP layer as a Base64 encoded string of the ASN.1 DER encoding of the certificate
```
```suggestion
<2> The CA certificate used to sign the node certificates that {es} uses for TLS on the HTTP layer.
The certificate is returned as a Base64 encoded string in ASN.1 DER encoding.
```

jkakavas replied:
I partially accepted this. The `ASN.1 DER` encoded version of the certificate _then_ gets Base64 encoded before we return it, your suggestion implies the opposite order

lockewritesdocs commented on x-pack/plugin/security/src/test/resources/org/elasticsearch/xpack/security/action/enrollment/README.asciidoc:
```diff
@@ -0,0 +1,44 @@
+== Instructions for gerating needed keystores
```
```suggestion
== Instructions for generating needed keystores
```

lockewritesdocs reviewed (commented):
Added some suggestions for updates to the Enroll Kibana API doc page.

jkakavas commented on x-pack/docs/en/rest-api/security/enroll-kibana.asciidoc:
```diff
@@ -0,0 +1,51 @@
+==== Example of enrolling a kibana instance
+
+[source,console]
+--------------------------------------------------
+GET /_security/enroll/kibana
+--------------------------------------------------
```
gotcha, thanks

jkakavas committed:
address docs feedback

jkakavas committed:
don't zerofill array

jkakavas commented:
@albertzaharovits I've addressed your feedback but 2 points ( getCertificate vs getCertificateChain  and bind vs publish address ) which need a little more discussion

jkakavas committed:
unused import

jkakavas commented:
@albertzaharovits do you want to take another look please and see if something needs to be resolved/discussed more or we should move forward with merging this? :pray:

jkakavas committed:
Merge remote-tracking branch 'origin/master' into enroll-client-api

jkakavas commented:
> @albertzaharovits do you want to take another look please and see if something needs to be resolved/discussed more or we should move forward with merging this? 

We went through remaining points in a zoom.

jkakavas committed:
Removed addresses from the Enroll Kibana API response as we determined Kibana doesn't need those

jkakavas committed:
update fingerprint as we're updating the CA certificate in httpCa.o12

jkakavas committed:
Merge remote-tracking branch 'origin/master' into enroll-client-api

jkakavas commented:
This change adds the Enroll Kibana API that enables a Kibana instance to
configure itself to communicate with a secured elasticsearch cluster
